Remote Joins server PR (#4888)
hasura/graphql-engine: docs/graphql/manual/api-reference/schema-metadata-api/remote-relationships.rst

.. meta::
   :description: Manage remote relationships with the Hasura schema/metadata API
   :keywords: hasura, docs, schema/metadata API, API reference, remote joins, remote relationships

Schema/Metadata API Reference: Remote Relationships
===================================================

.. contents:: Table of contents
  :backlinks: none
  :depth: 1
  :local:

> Remote Relationships will ensure independent remote schemas and related data types can be joined natively via Hasura.

Reviewer: ```suggestion
Remote Relationships allow you to join tables with remote schemas.
```

Let's keep it simple for now.

.. _create_remote_relationship:

create_remote_relationship
--------------------------

``create_remote_relationship`` is used to create a new remote relationship with an existing remote schema.

.. code-block:: http

   POST /v1/query HTTP/1.1
   Content-Type: application/json
   X-Hasura-Role: admin

  {
    "type": "create_remote_relationship",
    "args": {
      "name": "sample_remote_relationship",
      "table": "users",
      "hasura_fields": [
        "id"
      ],
      "remote_schema": "my-remote-schema",
      "remote_field": {

>         "messages": {

Reviewer: Will prefer a simpler `remote_field` definition here. Something which can be joined at top level.

Reviewer: Something like:

```
{
  "message": {
    "arguments": {
      "message_id":"$id"
       }
    }
}
 
```

          "arguments": {
            "where": {
              "id": {
                "eq": "$id"
              }
            }
          }
        }
      }
    }
  }

.. _create_remote_relationship_syntax:

Args syntax
^^^^^^^^^^^

.. list-table::
   :header-rows: 1

   * - Key
     - Required
     - Schema
     - Description
   * - name
     - true
     - RemoteRelationshipName_
     - Name of the remote relationship
   * - table
     - true
     - :ref:`QualifiedTable <QualifiedTable>`
     - Object with table name and schema
   * - hasura_fields
     - true
     - [:ref:`PGColumn <PGColumn>`]

>      - List of PG columns in the table that will be used in the ``remote_field`` object.

Reviewer: ```suggestion
     - Column(s) in the table that is used for joining with remote schema field. All join keys in ``remote_field`` must appear here.
```

I thought `hasura_fields` was removed as it was redundant. Should check.

Author: The `hasura_fields` are being used, I verified now.

   * - remote_schema
     - true
     - :ref:`RemoteSchemaName <RemoteSchemaName>`
     - Name of the remote schema to join with
   * - remote_field
     - true
     - RemoteField_

>      - Joining parameters of the remote join

Reviewer: ```suggestion
     - Name of the field in remote schema with the join parameters.
```

.. _update_remote_relationship:

update_remote_relationship
--------------------------

``update_remote_relationship`` is used to update an existing remote relationship.

.. code-block:: http

   POST /v1/query HTTP/1.1
   Content-Type: application/json
   X-Hasura-Role: admin

  {
    "type": "update_remote_relationship",
    "args": {
      "name": "sample_remote_relationship",
      "table": "users",

>       "hasura_fields": [],

Reviewer: This is not a good example for a remote relationship. Should change.

      "remote_schema": "my-remote-schema",
      "remote_field": {
        "messages": {
          "arguments": {
               "limit":10
            }
          }
        }
      }
    }
  }

.. _update_remote_relationship_syntax:

Args syntax
^^^^^^^^^^^

.. list-table::
   :header-rows: 1

   * - Key
     - Required
     - Schema
     - Description
   * - name
     - true
     - RemoteRelationshipName_
     - Name of the remote relationship
   * - table
     - true
     - :ref:`QualifiedTable <QualifiedTable>`
     - Object with table name and schema
   * - hasura_fields
     - true
     - [:ref:`PGColumn <PGColumn>`]
     - List of PG columns in the table that will be used in the ``remote_field`` object.
   * - remote_schema
     - true
     - :ref:`RemoteSchemaName <RemoteSchemaName>`
     - Name of the remote schema to join with
   * - remote_field
     - true
     - RemoteField_

>      - Joining parameters of the remote join

Reviewer: Pls change wrt to previous comment  in `create_remote_relationship`

.. _delete_remote_relationship:

delete_remote_relationship
--------------------------

``delete_remote_relationship`` is used to delete an existing remote relationship.

.. code-block:: http

   POST /v1/query HTTP/1.1
   Content-Type: application/json
   X-Hasura-Role: admin

   {
       "type" : "delete_remote_relationship",
       "args" : {
          "table":{
             "name":"users",
             "schema":"public"
          },
          "name":"sample_remote_relationship"
       }
   }

.. _delete_remote_relationship_syntax:

``delete_remote_relationship`` will delete an existing remote relationship.

Args syntax
^^^^^^^^^^^

.. list-table::
   :header-rows: 1

   * - Key
     - Required
     - Schema
     - Description
   * - table
     - true
     - :ref:`QualifiedTable <QualifiedTable>`
     - Object with table name and schema
   * - name
     - true
     - RemoteRelationshipName_
     - Name of the remote relationship

.. _RemoteRelationshipName:

RemoteRelationshipName
&&&&&&&&&&&&&&&&&&&&&&

.. parsed-literal::

  String

> RemoteField

Reviewer: You should specify a basic schema for this parameter. Something like:

```
FieldName: { 
  "arguments": InputArguments
}
```

&&&&&&&&&&&

.. code-block:: http

> 

Reviewer: Let's give 2 examples here. One being:

```
{
  "message": {
    "arguments": {
      "message_id":"$id"
       }
    }
}
```

   POST /v1/query HTTP/1.1
   Content-Type: application/json
   X-Hasura-Role: admin

   {
       "message": {
          "arguments":{
             "where": {
                 "id":{
                     "eq":"$id"
                  }
             }
          }
       }
   }

> 

Reviewer: This can be the second example as it is more complicated. Maybe you can show that you can add a static value join as well:

```
   {
       "messages": {
          "arguments":{
             "where": {
                 "id": {
                   "gt":"$id"
                  },
                 "limit": 100
             }
          }
       }
   }
```

Indentation should be fixed

.. meta::
   :description: Manage remote relationships with the Hasura schema/metadata API
   :keywords: hasura, docs, schema/metadata API, API reference, remote joins, remote relationships

Schema/Metadata API Reference: Remote Relationships
===================================================

.. contents:: Table of contents
  :backlinks: none
  :depth: 1
  :local:

> Remote Joins will ensure independent and related types in Remote Schemas can be joined natively via Hasura.

Reviewer 2: Is there a reason why we call it remote joins here and remote relationships in the rest of the page? We should probably be consistent with that.

Reviewer 2: Also, what does `independent` mean here? Can it be independent and related at the same time? :D

Author: @marionschleifer Yes, I will change "remote joins" to "remote relationships".

The independence here is referring to independent remote schemas and the related types here is referring to the data types of the join(For example:an Int field can only be joined to another Int field). 

It seems a bit confusing though, I will try to make it more clear

   * - name
     - true
     - RemoteRelationshipName_
     - Name of the remote relationship
   * - table
     - true
     - :ref:`QualifiedTable <QualifiedTable>`
     - Object with table name and schema
   * - hasura_fields
     - true
     - [:ref:`PGColumn <PGColumn>`]
     - Column(s) in the table that is used for joining with remote schema field. All join keys in ``remote_field`` must appear here.
   * - remote_schema
     - true
     - :ref:`RemoteSchemaName <RemoteSchemaName>`
     - Name of the remote schema to join with
   * - remote_field
     - true
     - RemoteField_
     - Name of the field in remote schema with the join arguments.

.. _delete_remote_relationship:

delete_remote_relationship
--------------------------

``delete_remote_relationship`` is used to delete an existing remote relationship.

.. code-block:: http

   POST /v1/query HTTP/1.1
   Content-Type: application/json
   X-Hasura-Role: admin

   {
       "type" : "delete_remote_relationship",
       "args" : {
          "table":{
             "name":"users",
             "schema":"public"
          },
          "name":"sample_remote_relationship"
       }
   }

.. _delete_remote_relationship_syntax:

> ``delete_remote_relationship`` will delete an existing remote relationship.

Reviewer 2: This we have twice now (once before the example)
